base/special/trig.jl

```diff
@@ -860,8 +860,10 @@ function cospi(x::T) where T<:Union{Integer,Rational}
     end
 end
 
-sinpi(x::Integer) = x >= 0 ? zero(float(x)) : -zero(float(x))
-cospi(x::Integer) = isodd(x) ? -one(float(x)) : one(float(x))
+sinpi(x::T) where {T<:Integer} = zero(signed(T))
```

Comment:
I think it's fine if `sinpi` and `cospi` have different return types, so that some computations remain `Unsigned` if all their underlying functions allow it. If someone plugs results of `sinpi` and `cospi` together, type promotion will fix that difference.

Reply:
Is the optimizer smart enough to remove the branch for toes where zero=-zero?

Reply:
I think I don't know what you mean. For integers there (usually?) is no difference between +0 and -0.

Reply:
I still think `sinpi` and `cospi` are too closely related to return different types. They are often used together, for example in `cispi` #35792 and `sincospi` #35816. Type promotion doesn't address all those problems, since the latter would produce an inhomogenous typed tuple and could produce quite unexpected and hard to debug results in edge cases. What would be the benefit of returning unsigned integers instead?

Reply:
It seems like type promotion as quiete preservative with `UInt`s, so I'm ok with your solution.
```
julia> typeof(UInt8(10) - Int8(2)) == typeof(UInt8(10) + Int8(2)) == UInt8
true
```

Reply:
@jonas-schulze, Julia assumes all platforms use two's complement integers (like current C++, unlike C and older C++), https://en.wikipedia.org/wiki/Ones%27_complement is very outdated, but given such a type (and with Julia you could make one, I'm not sure why you would), I'm not even sure if it should do the similar to floats:

```
julia> sin(-0.0)
-0.0

julia> sin(-0.0) = sin(-0)
ERROR: syntax: "-0" is not a valid function argument name around REPL[11]:1
Stacktrace:
 [1] top-level scope at REPL[11]:1

julia> sin(-0.0) == sin(-0)  # defined true, even with returning opposite signs (as -0 = 0):
true
```

Reply:
@PallHaraldsson That does sound like a very hypothetical example to me. I am not aware of Julia running on any non-two's-complement architectures. This assumption is also made throughout the whole code base, so if there was such a platform, these would probably be the least of our worries.

Reply:
Right, I was just addressing his "(usually?) is no difference between +0 and -0." I'm not worried about any of this, and would be ok with status quo, just like division returns floats (as natural at least there). Your change would work with all integer types, I know already defined, e.g. BitIntegers.jl, and all future types hopefully (like if someone made up one's complement on two-complements architecture).